joomla/joomla-cms
# JLanguage::loadLanguage performance improvements

#### Description

This PR tries to improve performance of `JLanguage::loadLanguage`.

It does so by replacing replacing the `array_merge` by `array_replace`.

No behaviour should be changed.
#### Performance

Since `JLanguage::loadLanguage` is used each time a ini language file is loaded, this PR adds a small performance boost, depending on if you're in the frontend or backend, the language your are using (en-GB or other), and the plugins/components/modules that you have that load language files in the page your viewing.
This small performance boost is for all Joomla pages (frontend, backend, offline, etc).

Some examples of the performance difference in the code changed:
###### PHP 5.6.16 (opcache enabled)

Sample site with multilanguage. Joomla without cache.

| Page | INI Files loaded | Before PR | After PR |
| --- | --: | --: | --: |
| Frontpage frontend (en-GB) | 11 | ~4.3 ms | ~1.6 ms |
| Frontpage frontend (other lang) | 22 | ~7.3 ms | ~2.1 ms |
| CPanel backend (en-GB) | 30 | ~16.2 ms | ~6.1 ms |
| CPanel backend (other lang) | 60 | ~34.7 ms | ~11.1 ms |
| Plugin list backend (en-GB) | 47 | ~24.9 ms | ~8.0 ms |
| Plugin list backend (other lang) | 94 | ~45.7 ms | ~14.4 ms |
###### PHP 5.3.10 (no opcache)

Ready to production site with multilanguage. Joomla without cache.

| Page | INI Files loaded | Before PR | After PR |
| --- | --: | --: | --: |
| Frontpage frontend (en-GB) | 21 | ~20.2 ms | ~7.6 ms |
| Frontpage frontend (other lang) | 42 | ~43.9 ms | ~15.9 ms |
| CPanel backend (en-GB) | 40 | ~123.2 ms | ~53.9 ms |
| CPanel backend (other lang) | 80 | ~202.1 ms | ~98.7 ms |
| Plugin list backend (en-GB) | 59 | ~123.7 ms | ~48.8 ms |
| Plugin list backend (other lang) | 118 | ~237.9 ms | ~91.2 ms |
#### How to test
1. Apply this patch.
2. Check in several languages if the language text is ok (test frontend and backend and multiple views).
3. Check also language overrides.
4. Check also if languages are ok in 3rd party extensions.
##### Testing Performance

Enable the debug and change the code in this PR (https://github.com/joomla/joomla-cms/blob/staging/libraries/joomla/language/language.php#L800-L812) to ...

Before PR:

``` php
JDEBUG ? JProfiler::getInstance('Application')->mark('- merging strings from ' . $filename) : null;
if (is_array($strings))
{
    // Sort the underlying heap by key values to optimize merging
    ksort($strings, SORT_STRING);
    $this->strings = array_merge($this->strings, $strings);
}

if (is_array($strings) && count($strings))
{
    // Do not bother with ksort here.  Since the originals were sorted, PHP will already have chosen the best heap.
    $this->strings = array_merge($this->strings, $this->override);
    $result = true;
}
JDEBUG ? JProfiler::getInstance('Application')->mark('- <strong>strings merged</strong>! JLanguage (instance ' . spl_object_hash($this) .') total strings: ' . count($this->strings)) : null;
```

After PR:

``` php
JDEBUG ? JProfiler::getInstance('Application')->mark('- merging strings from ' . $filename) : null;
if (is_array($strings) && count($strings))
{
    $this->strings = array_replace($this->strings, $strings, $this->override);  
    $result = true;
}
JDEBUG ? JProfiler::getInstance('Application')->mark('- <strong>strings merged</strong>! JLanguage (instance ' . spl_object_hash($this) .') total strings: ' . count($this->strings)) : null;
```

Reload the page and check the debugger Profile Information. (before and after PR). The ones that count are the time (in ms) of the "**strings merged**" messages in the profile information.

You will notice a slight improvement on each language ini load. Since there are a lot of language ini files, all together the performance improvement is the one i described in the table in this PR.
#### Observations

I hope i didn't forgot any test, but if anything wrong say so.
Suggestions, code reviews and improvements are welcome.
